velox/exec/tests/TopNRowNumberTest.cpp

```diff
@@ -27,14 +27,101 @@ namespace facebook::velox::exec {
+      .capturePlanNodeId(planNodeId)                      \
+      .nodename(partitionKeys, sortingKeys, limit, true)  \
+      .planNode()
+
 class TopNRowNumberTest : public OperatorTestBase {
  protected:
-  TopNRowNumberTest() {
+  explicit TopNRowNumberTest(core::TopNRowNumberNode::RankFunction function)
+      : function_(function) {}
+
+  void SetUp() override {
+    exec::test::OperatorTestBase::SetUp();
     filesystems::registerLocalFileSystem();
   }
+
+ protected:
+  const core::PlanNodePtr& topnNode(
+      PlanBuilder& planBuilder,
+      const std::vector<RowVectorPtr>& values,
+      const std::vector<std::string>& partitionKeys,
+      const std::vector<std::string>& sortingKeys,
+      int32_t limit,
+      bool generateRowNumber) {
+    switch (function_) {
+      case core::TopNRowNumberNode::RankFunction::kRowNumber:
+        return BUILD_TOPN(topNRowNumber);
+      case core::TopNRowNumberNode::RankFunction::kRank:
+        return BUILD_TOPN(topNRank);
+      case core::TopNRowNumberNode::RankFunction::kDenseRank:
+        return BUILD_TOPN(topNDenseRank);
+    }
+    VELOX_UNREACHABLE();
+  }
+
+  const core::PlanNodePtr& topnNodeId(
+      PlanBuilder& planBuilder,
+      core::PlanNodeId& planNodeId,
+      const std::vector<RowVectorPtr>& values,
+      const std::vector<std::string>& partitionKeys,
+      const std::vector<std::string>& sortingKeys,
+      int32_t limit,
+      bool generateRowNumber) {
+    switch (function_) {
+      case core::TopNRowNumberNode::RankFunction::kRowNumber:
+        return BUILD_TOPN_PLANNODEID(topNRowNumber);
+      case core::TopNRowNumberNode::RankFunction::kRank:
+        return BUILD_TOPN_PLANNODEID(topNRank);
+      case core::TopNRowNumberNode::RankFunction::kDenseRank:
+        return BUILD_TOPN_PLANNODEID(topNDenseRank);
+    }
+    VELOX_UNREACHABLE();
+  }
+
+  const core::PlanNodePtr& topnNodePartialFinal(
+      PlanBuilder& planBuilder,
+      core::PlanNodeId& planNodeId,
+      const std::vector<RowVectorPtr>& values,
+      const std::vector<std::string>& partitionKeys,
+      const std::vector<std::string>& sortingKeys,
+      int32_t limit) {
+    switch (function_) {
+      case core::TopNRowNumberNode::RankFunction::kRowNumber:
+        return BUILD_TOPN_PARTIAL_FINAL(topNRowNumber);
+      case core::TopNRowNumberNode::RankFunction::kRank:
+        return BUILD_TOPN_PARTIAL_FINAL(topNRank);
+      case core::TopNRowNumberNode::RankFunction::kDenseRank:
+        return BUILD_TOPN_PARTIAL_FINAL(topNDenseRank);
+    }
+    VELOX_UNREACHABLE();
+  }
+
+  const core::TopNRowNumberNode::RankFunction function_;
 };
 
-TEST_F(TopNRowNumberTest, basic) {
+class MultiTopNRowNumberTest : public TopNRowNumberTest,
+                               public testing::WithParamInterface<
+                                   core::TopNRowNumberNode::RankFunction> {
+ public:
+  MultiTopNRowNumberTest() : TopNRowNumberTest(GetParam()) {}
+};
+
```

Comment:
Could you also add a test case for the logic in fixTopRank.

Reply:
@liujiayi771 : The fixTopRank logic is tested very thoroughly in the fewPartitions test. Have added a comment there.